lib/common/tests/schemas/pcmk__schemas_later_than_test.c

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2023 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU General Public License version 2
+ * or later (GPLv2+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+
+#include <crm/common/unittest_internal.h>
+#include <crm/common/lists_internal.h>
+
+#include <glib.h>
+
+static int
+setup(void **state) {
+    setenv("PCMK_schema_directory", G_STRINGIFY(PCMK__TEST_SCHEMA_DIR), 1);
+    crm_schema_init();
+    return 0;
+}
+
+static int
+teardown(void **state) {
+    crm_schema_cleanup();
+    unsetenv("PCMK_schema_directory");
+    return 0;
+}
+
+static void
+invalid_name(void **state)
+{
+    assert_null(pcmk__schemas_later_than("xyz"));
+    assert_null(pcmk__schemas_later_than("pacemaker-"));
+}
+
+/* The list of known schemas will change every time we make a significant update,
+ * and we don't want to have to change this test every single time.  This means a
+ * couple things:
+ *
+ * (1) We can really only test that there's AT LEAST as many elements in the list
+ *     as we expect.  We can't test the exact number.
+ *
+ * (2) We can't test the contents of the list as well as we might want.
+ *     pcmk__schemas_later_than returns a list with the most recent version sorted
+ *     first, and we don't know what the most recent version will be.  Therefore we
+ *     can only check that the end of the list is as expected.
+ */
+static void
+valid_name(void **state)
+{
+    GList *schemas = NULL;
+
+    schemas = pcmk__schemas_later_than("pacemaker-1.0");
+    assert_true(g_list_length(schemas) >= 23);
+    /* There is no "pacemaker-1.1". */
+    assert_string_equal("pacemaker-1.2", g_list_nth_data(schemas, 0));
+    g_list_free_full(schemas, free);
+
+    schemas = pcmk__schemas_later_than("pacemaker-2.0");
+    assert_true(g_list_length(schemas) >= 20);
+    assert_string_equal("pacemaker-2.1", g_list_nth_data(schemas, 0));
+    g_list_free_full(schemas, free);
+
+    /* Check that "pacemaker-2.10" counts as later than "pacemaker-2.9". */
+    schemas = pcmk__schemas_later_than("pacemaker-2.9");
+    assert_true(g_list_length(schemas) >= 11);
+    assert_string_equal("pacemaker-2.10", g_list_nth_data(schemas, 0));
+    g_list_free_full(schemas, free);
+
+    schemas = pcmk__schemas_later_than("pacemaker-3.0");
+    assert_true(g_list_length(schemas) >= 9);
+    assert_string_equal("pacemaker-3.1", g_list_nth_data(schemas, 0));
+    g_list_free_full(schemas, free);
+
+    /* And then something way in the future.  Hopefully this is far enough
+     * out to where we don't have to change this test any time soon.
+     */
+    schemas = pcmk__schemas_later_than("pacemaker-9.0");
```

Comment:
999 to be really safe :)

Alternatively, create a test schema directory with symlinks to a few schemas and use that

Reply:
Yeah, I could create lib/common/tests/schemas/schemas, add a couple symlinks in there, and then set PCMK__TEST_SCHEMA_DIR to that.  This could probably also all be done at build time, without putting the symlinks in git.  For instance, I could create the directory and symlink over the ten oldest schema files.  Then I'd know what to expect and we wouldn't be chasing whatever happens with the newest ones.  I think we'd still be okay from a thoroughness of testing perspective.

Reply:
Updated to add this.